### DISCUSSION TITLE
Update multiple rows to table

### DISCUSSION THREAD

**Discussion by applicationuser** [Budibase/budibase#8432] [Help]
Hello,

Need your help on how to design a form with multiple data rows update to a single table.

Data from budibase table displayed using repeater form. Based on the user selection criteria , repeater form can have at a minimum of 2 rows to 20 rows of data displayed.  similar to a inspection/survey form with list of questions to answer & rating

Layout .
-->Form
----->Data Provider
---------> Repeater
------------->Field Group

is there an option to place save button outside of repeater and trigger update of all the rows selected from Data provider ?

prefer not to have Data provider limit to 1 and save each row, that would require user to go through ~200 rows and save each row.

Thanks,

**Comment by melohagan** [comment]
Hey @applicationuser 

Part of the challenge is that the form fields within a repeater cannot have unique identifiers. Something noted here: https://github.com/Budibase/budibase/issues/5944

This can be alleviated by repeating the form as well, so that your component tree looks as follows:
![Screenshot 2022-11-01 at 10 31 58](https://user-images.githubusercontent.com/101575380/199214157-ba272050-00ef-4ad9-bd99-bd0c0a25c7de.png)

The tricky bit then is to aggregate the forms data outside the form component scope. The way I can think to do this is to add an [Update State](https://docs.budibase.com/docs/app-state) action to the **On Change** event of each form field (this could be a bit tedious if there are many form fields unfortunately).

Then you can update a JS dictionary to keep track of each form state, using the repeater row id as the index, e.g.
```js
let formState = $("State.FormState");
const index = $("New Repeater.Sales.Auto ID")-1;

if (!formState) {
	formState = {};
}
if (!formState[index]) {
	formState[index] = {};
}
formState[index]["itemType"] = $("Field Value");
formState[index]["_id"] = $("New Repeater.Sales._id");

return formState;
```

Next we need a button that will trigger an automation. We need to do this because automations support looping over rows, whereas actions client side do not. 

![Screenshot 2022-11-01 at 11 54 58](https://user-images.githubusercontent.com/101575380/199227118-9ed571bf-2e6d-42f2-aeb2-c3ad7fdc6d18.png)

```js
return $("State.FormState") ? Object.values($("State.FormState")) : [];
```
*This JavaScript takes our dictionary and converts it to an array so that the automation can loop over it*

#### Automation
![Screenshot 2022-11-01 at 11 56 05](https://user-images.githubusercontent.com/101575380/199227295-da5c0565-e222-44f5-ad16-e7b9c584f321.png)

![Screenshot 2022-11-01 at 11 56 25](https://user-images.githubusercontent.com/101575380/199227342-10579b59-f90b-4878-a2f1-6c5881468007.png)
*I've only included some of the fields as an example*

Here is the app export so you can take a closer look: 
[Multi Row Auto-export-1667303882919.tar.gz](https://github.com/Budibase/budibase/files/9909518/Multi.Row.Auto-export-1667303882919.tar.gz)

**Reply by applicationuser** [reply]
Thank You @melohagan for putting this together and providing app export. 

i'll work through with our form data and let you know if anything else needed.  Thanks again!

**Reply by melohagan** [reply]
You're welcome! 😃

**Comment by gmentzik** [comment]
HI @melohagan ,
 I am trying to create a similar form. Unfortunately I noticed that the onChange returns the old field value while building the json object at state variable. So something is wrong here. I am on version: 2.1.31
![image](https://user-images.githubusercontent.com/65217733/203765465-24461ed8-1f5b-4596-a3cb-8ce32195c48a.png)

**Reply by melohagan** [reply]
Hey @gmentzik 

Are you using the `{{ Field Value }}` binding? 

![Screenshot 2022-11-24 at 10 49 47](https://user-images.githubusercontent.com/101575380/203765774-6c7e5e65-33bc-4299-9ea9-901d40dcdef2.png)

![Screenshot 2022-11-24 at 10 50 05](https://user-images.githubusercontent.com/101575380/203765824-887db33f-a1fa-47f5-9c92-d23684ebb0dc.png)

**Reply by gmentzik** [reply]
No was using Form.Field... 
formState[settingID]["pmiValue"] = $("PMIForm.Fields.pmiValue");
Changed to Field Value
formState[settingID]["pmiValue"] = $("Field Value");

Now it has new value! Thank You!
![image](https://user-images.githubusercontent.com/65217733/203769941-fbabd598-a642-4beb-82be-b09991eff0cf.png)

**Comment by gmentzik** [comment]
A few comments here as I am also trying to use looping to update/insert multiple rows but I am not very satisfied. Here are my comments: 
1. Update works for both updating and creating row (if _id is missing). That simplifies implementation otherwise you would need to create two arrays (insert and update).
2. I was not able to test the automation from backend using Finish and test. I am getting an error that my id field in database that is set to Not Null should have a default value. When testing from Frontend row is stored/updated without issue.
3. Automation only works from published app which is a pity as in such cases preview is very valuable.
4. Failing insert/updates are not propagated to FE. So user does not know if updates were performed unless he refreshes screen or try to refresh provider to check by himself.
5. Hint. You may need to cleanup your state after submitting to automation.

**Comment by LMPexplorer** [comment]
Hi @melohagan, 

I am using version : 2.19.4.
I followed your tuto to be able to update multiple rows in a table using update state/ automation. 
I used it for two different tasks. 
One is working but unfortunately, the other one doesn't seem to work. the field value part is working, but instead of updating an existing row, it create a new row, so my "_id" part doesn't seem to work. I have thoroughly followed ur instructions on both tasks so i don't understand why the second task isn't working.
Here is screenshot of the second task : 

![image](https://github.com/Budibase/budibase/assets/158468320/7c79edf2-aca2-4315-9ce6-ecc565de0cd7)
![image](https://github.com/Budibase/budibase/assets/158468320/c104fc3e-b981-4792-9702-64c5f0fbd16d)
![image](https://github.com/Budibase/budibase/assets/158468320/39fa404a-87a6-48dc-9640-21a266ace24c)
![image](https://github.com/Budibase/budibase/assets/158468320/547e359b-9718-4043-81f8-c524474b0cc6)
![image](https://github.com/Budibase/budibase/assets/158468320/1e0ec1f5-4b8b-4ef1-8906-3cf4085eeafe)
![image](https://github.com/Budibase/budibase/assets/158468320/ab0645e7-fe6f-4535-af0d-f4e2663a385f)

Do you find any error ? 
Thank you

**Reply by melohagan** [reply]
Step 1: Update to the latest version 😄  **2.22.18**

**Reply by LMPexplorer** [reply]
Hi @melohagan, thank you for your quick response. I have updated the app to the latest version. Unfortunately, it doesn't seem to fix my error. It still create a new row instead of updating an existing row. Do you find any misconduct that could fix my error ?

**Reply by LMPexplorer** [reply]
I have finally found my error. But i see that sometimes when a value is updated, the automation doesn't trigger. Could you explain to me how it works ? Is there a way to make sure that the automation trigger when a value is updated ? Thank you :)

**Reply by melohagan** [reply]
If you call the **Trigger Automation** action from the Design section, then it should always trigger the automation. Maybe you are missing some input data in those cases it doesn't work as you would expect?

**Reply by LMPexplorer** [reply]
Thank you for your quick response. I am using a repeater and the action is in the "on change" of the repeater so it should be working for all my fields. I will see what i am missing. Thank you for your help !

**Reply by phifogg** [reply]
Thanks so much for this tutorial! Works like charm!
